File: shoring/compat/src/main/java/com/panopset/compat/Zombie.java
package com.panopset.compat;

import java.util.ArrayList;
import java.util.List;

public class Zombie {

  private boolean active = true;

  public void stop() {
    if (!isActive()) {
      return;
    }
    active = false;
    for (Runnable sa : stopActions) {
      sa.run();
    }
  }

  public void resume() {
    active = true;
  }

  public boolean isActive() {
    return active;
  }

  public void addStopAction(Runnable runnable) {
    if (!stopActions.contains(runnable)) {
      stopActions.add(runnable);
    }
  }

  public void removeStopAction(Runnable runnable) {
    stopActions.remove(runnable);
  }

  private final List<Runnable> stopActions = new ArrayList<>();
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/Zombie.kt
package com.panopset.compat

object Zombie {
    var isActive = true
        private set

    fun stop() {
        if (!isActive) {
            return
        }
        isActive = false
        for (sa in stopActions) {
            sa.run()
        }
    }

    fun resume() {
        isActive = true
    }

    fun addStopAction(runnable: Runnable) {
        if (!stopActions.contains(runnable)) {
            stopActions.add(runnable)
        }
    }

    fun removeStopAction(runnable: Runnable) {
        stopActions.remove(runnable)
    }

    private val stopActions: MutableList<Runnable> = ArrayList()
}
